### PR TITLE
Update Rust to Nightly 2017-01-12

### PR DESCRIPTION
**Pull request by Hoverbear**
Prior to this change, we used 2017-08-09.

This change moves us to 2018-01-12.

Contributes to #2615.
 
Depends on https://github.com/pingcap/SRE/pull/63.

### CODE REVIEW

**Comment by breeswish** [comment]
I believe that some dependencies in cargo needs update as well since several weeks ago I failed to use the latest rust nightly to compile TiKV.

**Comment by Hoverbear** [comment]
There are some new clippy warnings to fix. :)

**Comment by Hoverbear** [comment]
Fun fact: You can suppress all warnings while fixing specific lints like this:

1. Pick a lint, and `#![deny($YOUR_LINT)]` in `src/lib.rs`.
2. Run `RUSTFLAGS=-Awarnings rustup run nightly cargo clippy`
3. Fix them all.
4. Remove the deny, continue to the next.

**Comment by siddontang** [comment]
@Hoverbear, 
I think using `clone` directly looks neater than `Arc::clone`, but it can't pass the lint now. 😭

**Comment by siddontang** [comment]
Hi @Hoverbear 

Can we choose a new version and go one this?

**Comment by Hoverbear** [comment]
I largely agree with you but it does improve understanding some though.

I'm going to bump to a newer version yes. :)

File: README.md [hunk 1]
```diff
@@ -23,7 +23,15 @@ Thanks to the internal optimization, TiKV and TiDB can work together to be the b
 
 ### Required Rust version
 
-Rust Nightly is required.
+Rust Nightly is required. TiKV is currently tested mainly with `rust-nightly-2018-01-09`, however we would like to track `nightly`, so please report new breakage.
+
+```bash
+# Get rustup from rustup.rs, then:
+rustup install nightly
+cargo +nightly install rustfmt-nightly
```
Comment (siddontang) [inline README.md]: do we need to add vers 0.3.4 here?

File: README.md [hunk 2]
```diff
@@ -23,7 +23,15 @@ Thanks to the internal optimization, TiKV and TiDB can work together to be the b
 
 ### Required Rust version
 
-Rust Nightly is required.
+Rust Nightly is required. TiKV is currently tested mainly with `rust-nightly-2018-01-09`, however we would like to track `nightly`, so please report new breakage.
+
+```bash
+# Get rustup from rustup.rs, then:
+rustup install nightly
```
Comment (siddontang) [inline README.md]: do we need to assign a version here?

Comment (Hoverbear) [reply README.md]: If we're trying to track nightly in development, then no. If we'd like development to lock to a version and update it in jumps then we do. I'm not sure what we'd prefer.

Comment (siddontang) [reply README.md]: I think we should use the specified version here. We meet many times that the user uses a newer Rust version and fails to compile TiKV.

**Comment by Hoverbear** [comment]
🎉 Circle CI passes.

**Comment by ngaut** [comment]
Cool, please keep going.

**Comment by siddontang** [comment]
Our Jenkins CI uses old rust version, so Jenkins fails

**Comment by Hoverbear** [comment]
@siddontang https://github.com/pingcap/SRE/pull/63 should fix this.

**Comment by Hoverbear** [comment]
Hm... http://45.249.247.133:8080/blue/organizations/jenkins/tikv_ghpr_build/detail/tikv_ghpr_build/1968/pipeline says it's still getting 2017-08-09. 🤔

**Comment by siddontang** [comment]
PTAL @iamxy

**Comment by iamxy** [comment]
CI updated

**Comment by Hoverbear** [comment]
I think I found the missing settings: https://github.com/pingcap/SRE/pull/64

**Comment by iamxy** [comment]
As we discussed, could you add a file named `RUST_VERSION` which content like this:
```
# This file is used for specifying Rust version for Jenkins CI
nightly-2018-01-12
```

I have fixed the CI to handle it .

**Comment by ngaut** [comment]
Amazing. PTAL @siddontang @zhangjinpeng1987 @overvenus

**Comment by siddontang** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @BusyJay

**Comment by Hoverbear** [comment]
I rebased since @overvenus 's PR got merged. 🎉

File: RUST_VERSION [hunk 3]
```diff
@@ -0,0 +1,2 @@
+# This file is used for specifying Rust version for Jenkins CI
+nightly-2018-01-12
```
Comment (BusyJay) [inline RUST_VERSION]: Can circle ci reuse this?

**Comment by Hoverbear** [comment]
🎉

File: src/coprocessor/codec/mysql/decimal.rs [hunk 4]
```diff
@@ -1759,18 +1748,19 @@ macro_rules! read_word {
...
+                        | u32::from(buf[2])
                 } else {
-                    ((buf[0] as u32) << 16) | ((buf[1] as u32) << 8) | (buf[2] as u32)
+                    (u32::from(buf[0]) << 16) | (u32::from(buf[1]) << 8) | u32::from(buf[2])
                 }
             }
             4 => {
-                (((buf[0] as i8 as i32) << 24) + ((buf[1] as i32) << 16) +
-                 ((buf[2] as i32) << 8) + (buf[3] as i32)) as u32
+                ((i32::from(u8::from(buf[0])) << 24) + (i32::from(buf[1]) << 16) +
```
Comment (BusyJay) [inline src/coprocessor/codec/mysql/decimal.rs]: `u8`?

Comment (Hoverbear) [reply src/coprocessor/codec/mysql/decimal.rs]: Oh no, you're right.

File: RUST_VERSION (hunk 3 above)
Comment (Hoverbear) [reply RUST_VERSION]: I think that's a great idea. Might be good to have the rustfmt version as well somehow.

File: src/coprocessor/codec/mysql/decimal.rs (hunk 4 above)
Comment (Hoverbear) [reply src/coprocessor/codec/mysql/decimal.rs]: Hm, fixing that gives an error... See #2693.

File: RUST_VERSION (hunk 3 above)
Comment (BusyJay) [reply RUST_VERSION]: If rustfmt is installed via `rustup component`, then version is unnecessary anymore.

**Pull request merged**